utils/oscap-remediate-offline

```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env bash
+
+# SPDX-License-Identifier: GPL-2.0-or-later
+# Copyright (C) 2021 Red Hat Inc., Durham, North Carolina.
+#
+# Evgenii Kolesnikov <ekolesni@redhat.com>
+#
+# The setup script for the offline system remediation (systemd's /system-update) service
+
+SYSTEMD_SYSTEM_UPDATE_LINK="/system-update"
+OSCAP_REMEDIATE_CONF="/var/tmp/oscap-remediate-offline.conf.sh"
+
+die() {
+    echo "$*" >&2
+    exit 1
+}
+
+invalid() {
+    echo -e "$*\n" >&2
+    usage
+    exit 1
+}
+
+usage() {
+    echo "oscap-remediate-offline -- Setup script for the SCAP evaluation and remediation using systemd'd update service."
+    echo
+    echo "Usage:"
+    echo
+    echo "$ oscap-remediate-offline PROFILE INPUT_CONTENT"
```

Comment:
If you take a look at usage of OpenSCAP in OAA e.g. near https://github.com/OpenSCAP/oscap-anaconda-addon/blob/rhel8-branch/org_fedora_oscap/common.py#L259, then more arguments will be likely needed:
- tailoring file,
- datastream ID, XCCDF ID,
- path for results, report

Although you provide a meaningful default in your script, it makes sense that the addon should be able control the location where reports are generated.

If you don't like the idea of supporting multiple arguments, some of them optional, then [Argbash](https://argbash.readthedocs.io/en/latest/) may help - wink wink.

Reply:
Actually the interface between OpenSCAP and Image Builder or OAA is the configuration file along with the symlink `/system-update` pointing to it. The `oscap-remediate-offline` was never meant to be used by them because it would definitely be of no help to IB, and I suspect for Anaconda it would be easier to create a symlink in a proper place rather thank juggle with the chroot environment and try executing the script.

It is just a convenient way of testing the thing. Now I'm not even sure it is a good idea to advertise it (via man page or the user manual).

With that being said, extending the configuration with tailoring, DS and XCCDF ids and a path for the result is definitely a good idea. The question is, what should be optional:
- ds file name (definitely required)
- profile_id
- ds_id
- xccdf_id
- path for logs (definitely optional)
- path for results (should it be separated from the logs)?

Reply:
I think that having a ready-to-use way of scheduling the remediation for the next boot makes sense. If I understand it right, a tool would have to make those substitutions in the config file, copy stuff in place, and create a symlink - looks like a good job for a script.

Regarding which args should be optional - probably all except datastream and profile ID.
And what do you mean by logs? The scanner produces a report and results/ARF, I don't see it capturing logs. Anyway, I think that the option to specify independently where to save those or that we don't want them (probably by providing an empty filename if the default is to produce them somewhere) again makes sense.

Reply:
While it may seem convenient to have this script, its introduction bears some implications.

First of all it defies declarative over imperative approach we are trying to employ. It also creates a dependency on a bash script (and also the bash itself) and the necessity of executing something at the configuration (priming) step. And I would rather put some effort to discourage usage of this script beyond testing than to make it more versatile.

Moreover, it is an API. And we would be expanding API surface without a good reason (if we no longer consider it as an interface for OAA or IB).

As for the value for the end-user, I can't see any benefits in choosing systemd service over a regular scan and remediation. If nothing else it would move the process into less controllable and observable environment.

Reply:
I can see the implication of expanding the API, the script would come up with an API of a higher level of abstraction. We have `oscap-podman`, `oscap-chroot` and so on that do exactly this - they hide implementation details of various execution modes of the scanner. I don't see any concern with expanding our "API surface", as the script is a facade that don't get complicated. Is there something that you have in mimd what could go wrong?
I would also argue that the addon shouldn't have such low-level knowledge of how to schedule the firstboot scan. But I also disagree with the imperative vs declarative claim - although Bash is not a declarative language, the script's interface is the closest to declarative that I can imagine - provided arguments define the result, and you don't care what needs to be done to get that result - isn't that what declarative is supposed to mean?

Reply:
While I found the script very convenient for performing my review, I slightly lean towards removing the script. 

To make it working properly we would have to improve the `oscap info` tool to be able to verify that the provided parameters are correct and will lead to a meaningful remediation. Improving `oscap info` wouldn't be simple. 

Keep in mind that in imagebuilder we can't execute any bash script, so we will have to declare the config file options as a public API anyway. So the script doesn't help us with the abstraction that we would want, it doesn't allow us to fully hide implementation details from the outer world.

Reply:
I don't have a strong opinion about the script, the advantage of the script I see is that whenever we change something in API we won't need to update OAA component. And regarding the argument about validating parameters of the script, I think we could do just some basic sanity check and let it fail during boot in case something is invalid, user will see it failed from the logs - that would be the same as if one does some mistake when writing parameters directly into a config file.

Also as mentioned the API needs to be extended to accept more parameters, I think what @evgenyz listed looks good:
```
Required:
ds file name
profile_id

Optional:
ds_id
xccdf_id
tailoring
path for XML results
path for HTML report
path for logs
```